### ISSUE TITLE
Issues with ifconfig, sockets, nmap, ping, nslookup, dig and others not working

### ISSUE THREAD

**Issue by javydekoning** [Microsoft/BashOnWindows#69]
nmap:
<code>
root@localhost:/mnt/c/users/jdekoning# nmap google.com

Starting Nmap 6.40 ( http://nmap.org ) at 2016-04-08 12:37 UTC
route_dst_netlink: cannot create AF_NETLINK socket: Invalid argument
</code>

Ping:
<code>
root@localhost:/mnt/c/users/jdekoning# ping google.com
ping: icmp open socket: Socket type not supported
</code>

nslookup:
<code>
root@localhost:/mnt/c/users/jdekoning# nslookup google.com
socket.c:2447: setsockopt(20, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 10.142.23.200#53: Invalid argument
socket.c:2447: setsockopt(21, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 10.141.23.200#53: Invalid argument
socket.c:2447: setsockopt(22, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 10.130.23.100#53: Invalid argument
</code>

traceroute:
<code>
root@localhost:/mnt/c/users/jdekoning# traceroute 8.8.4.4
traceroute to 8.8.4.4 (8.8.4.4), 30 hops max, 60 byte packets
setsockopt IP_MTU_DISCOVER: Invalid argument
</code>

DIG: 
<code>
root@localhost:/mnt/c/users/jdekoning# dig @8.8.4.4 google.com
socket.c:2447: setsockopt(20, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 8.8.4.4#53: Invalid argument
</code>

**Comment by Aneurin** [comment]
All of the ones which produce "socket.c: 1915: internal_send ..." are BIND tools that don't seem to work here. You can install the Knot DNS equivalents which do, using `apt-get install knot-host knot-dnsutils`, to get working `dig` and `host` at least, though it doesn't seem like there's a Knot `nslookup`.

For the rest, it looks like only certain types of sockets can be created. Maybe raw sockets aren't permitted so we can only open TCP and UDP sockets? You would at least expect ICMP to be available though.

**Comment by sunilmut** [comment]
Yes, NETLINK sockets are currently not implemented. ICMP is also currently not supported. UDP and TCP sockets should work fine though. We will be tracking\prioritizing which features to light up next by looking at the feedback provided at the wpdev portal. We appreciate your feedback and please make sure to provide it through the wpdev portal as well. Looks like there is already a post tracking this which you can up vote (or add a new one):

https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13383789-enable-network-interface-enumeraration

**Comment by xiaoyu9964** [comment]
so when this problem can be solved?

**Comment by sunilmut** [comment]
@xiaoyu9964 Thanks for your patience and we understand that not having basic network tools working,  can be frustrating. That being said, support for RAW/ICMPv4 sockets are being checked in our development branch. That should unblock ping. ping6 still doesn't work. We are working to unblock these scenarios as well in future. Meanwhile, please help us prioritize this by up voting this specific request in the [User Voice page] (https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13383789-enable-network-interface-enumeraration). Thanks again for trying WSL and providing feedback.

**Comment by MadcowD** [comment]
This issue is causing issues with libzmq and other socket libraries.

Related issue is #134

**Comment by Deshke** [comment]
any update on this? network is a rather big point and should work

**Comment by aseering** [comment]
A lot has changed since this ticket was last updated.  It's probably be worth re-testing the specific issues reported here.

**Comment by ldanicel** [comment]
I just fresh installed 4 days ago and it acts the same.

**Comment by aseering** [comment]
This issue is actually an aggregation of a bunch of bugs, some of which are known to be fixed.  Could you please include your specific Windows build and which of the issues on this ticket you tested?

**Comment by aseering** [comment]
In particular, off the top of my head, ping is #18 and libzmq is #185.

Note that both are fixed with caveats; more details are available on those tickets.

**Comment by ldanicel** [comment]
traceroute, mtr, ifconfig, ip, route, tcpdump, nmap are not functioning at all
`
root@CASH:~# mtr google.com
mtr: unable to get raw sockets.

root@CASH:~# traceroute google.com
traceroute to google.com (82.76.231.39), 30 hops max, 60 byte packets
setsockopt IP_MTU_DISCOVER: Invalid argument 

root@CASH:~# nmap -O google.com
Starting Nmap 6.40 ( http://nmap.org ) at 2016-07-18 00:13 DST
route_dst_netlink: cannot bind AF_NETLINK socket: Invalid argument

root@CASH:~# route -n
/proc/net/route: No such file or directory
INET (IPv4) not configured in this system.


root@CASH:~# ip addr show
SO_SNDBUF: Invalid argument

root@CASH:~# tcpdump
tcpdump: getifaddrs: Invalid argument

`

ping looks to be ok but still has some spookie latencies reported, like:

`root@CASH:~# ping google.com
PING google.com (82.76.231.25) 56(84) bytes of data.
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=1 ttl=61 time=0.518 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=2 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=3 ttl=61 time=2.52 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=4 ttl=61 time=2.36 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=5 ttl=61 time=2.00 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=6 ttl=61 time=2.50 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=7 ttl=61 time=2.53 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=8 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=9 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=10 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=11 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=12 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=13 ttl=61 time=0.000 ms
64 bytes from 82-76-231-25.rdsnet.ro (82.76.231.25): icmp_seq=14 ttl=61 time=2.41 ms
`

windows build is Microsoft Windows [Version 10.0.14388]

**Comment by Deshke** [comment]
i'm currently on Windows 10 Insider Preview 14390 - and i've got the same issues as in https://github.com/Microsoft/BashOnWindows/issues/69#issuecomment-233204632

**Comment by Skalin** [comment]
14394 - still not working properly, traceroute does not work

**Comment by vnodecg** [comment]
Any update on this?

**Comment by CharlieEriksen** [comment]
When you started Bash with administrative privileges, not sudo, and nmap 192.168.1.1, you get this error instead:

```
Starting Nmap 6.40 ( http://nmap.org ) at 2016-08-04 19:39 DST
Problem binding to interface , errno: 22
socket_bindtodevice: Invalid argument
NSOCK ERROR [1.1340s] mksock_bind_device(): Setting of SO_BINDTODEVICE failed (IOD #1): Invalid argument (22)
NSOCK ERROR [1.1340s] mksock_bind_device(): Setting of SO_BINDTODEVICE failed (IOD #2): Invalid argument (22)
NSOCK ERROR [1.1350s] mksock_bind_device(): Setting of SO_BINDTODEVICE failed (IOD #3): Invalid argument (22)
Problem binding to interface , errno: 22
socket_bindtodevice: Invalid argument
Problem binding to interface , errno: 22
socket_bindtodevice: Invalid argument
Problem binding to interface , errno: 22
```

Ping works when you run it with administrative privileges.

**Comment by DennisSc** [comment]
inetutils-traceroute gives the following error:
traceroute: setsockopt: invalid argument

ping works now, and name resolution too (I added 127.0.0.1 mywindowspcname. to /etc/hosts, don't know if this helps but it remedied the error "host not found" when issuing pings to IP addresses I think)

Lokks like things are still under construction...

**Comment by sunilmut** [comment]
@DennisSc - Yes, we are aware of the pain of not having some of these networking tools enabled on WSL and are looking into it. Please also help us prioritize by upvoting the [user voice page for this feature] (Here is the user voice page for supporting iptables, ifconfig and others!
https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/15202875-would-you-like-to-support-ifconfig-iw-iwconfig-t)

**Comment by sunilmut** [comment]
https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/15202875-would-you-like-to-support-ifconfig-iw-iwconfig-t

**Comment by Honey0cool** [comment]
root@xxxx:~# ifconfig
Warning: cannot open /proc/net/dev (No such file or directory). Limited output.
root@xxxx:~# ping www.google.com
ping: icmp open socket: Permission denied
root@xxxx:~# ip r
SO_SNDBUF: Invalid argument
root@xxxx:~# cat /etc/hosts
127.0.0.1 localhost

# The following lines are desirable for IPv6 capable hosts
::1 ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
ff02::3 ip6-allhosts
root@xxxx:~# cat /etc/resolv.conf
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 192.168.1.1
nameserver 172.18.11.1
nameserver 192.168.1.1

----------------------------
After Reinstall Windows 10’s Ubuntu Bash Shell

lxrun /uninstall /full
lxrun /install /y

Now “root” Account as the default user account . What ?

**Comment by insanikov** [comment]
@Honey0cool You shouldn't run `lxrun /install` as administrator, that's what causes "root" to be the default account.

**Comment by Honey0cool** [comment]
@insanikov Then What commands i Need To Use, "lxrun /install" . not  "lxrun /install /y" Correct. 
What About Ifconfig not Working.

**Comment by bitcrazed** [comment]
@Honey0cool I recommend NOT running `lxrun /install /y` as this bypasses the user account creation step resulting in you logging in as root.

Instead, run `lxrun /install` (omitting the `/y`) which will first ask you to create a UNIX user account and then you'll run as that standard account by default, and use `sudo` to elevate when necessary.

**Comment by aseering** [comment]
Random observation:  There seem to be a bunch of scattered tickets and UserVoices for this issue.  There's a UserVoice ticket about this issue that says "this is the wrong forum, please file it in Github"; there are Github tickets that refer people to the UserVoice.  (As a blanket statement; not just to upvote the existing technical issue.)

I think it would be helpful if a forum mod could go through both forums, close out dupes, and post comments telling us which ticket should be used where for what.  A little clean-up, so to speak.

**Comment by gudenuff** [comment]
I am having issues creating network namespace using _netns_

## Steps to repro
Execute the following command from a privileged user
`ip netns add <namespace_name>`
where <namespace_name> is the desired network namespace.
## Observed behavior
An error returned while trying to set socket options:
`setsockopt(3, SOL_SOCKET, SO_SNDBUF, [32768], 4) = -1 EINVAL (Invalid argument)`
## Example output

```
user@computer:~$ sudo ip netns add vpn
sudo: unable to resolve host COMPUTER
[sudo] password for user:
SO_SNDBUF: Invalid argument
```

## Detailed tracing
Here's the more detailed tracing produced by _strace_
```
root@computer:/home/user# strace ip netns add vpn
execve("/sbin/ip", ["ip", "netns", "add", "vpn"], [/* 16 vars */]) = 0
brk(0)                                  = 0xe6b000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f8464390000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=34552, ...}) = 0
mmap(NULL, 34552, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f8464387000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320\16\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=14664, ...}) = 0
mmap(NULL, 2109744, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f8463df0000
mprotect(0x7f8463df3000, 2093056, PROT_NONE) = 0
mmap(0x7f8463ff2000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x2000) = 0x7f8463ff2000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0P \2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1840928, ...}) = 0
mmap(NULL, 3949248, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f8463a20000
mprotect(0x7f8463bda000, 2097152, PROT_NONE) = 0
mmap(0x7f8463dda000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1ba000) = 0x7f8463dda000
mmap(0x7f8463de0000, 17088, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f8463de0000
close(3)                                = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f8464380000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f8464370000
arch_prctl(ARCH_SET_FS, 0x7f8464370740) = 0
mprotect(0x7f8463dda000, 16384, PROT_READ) = 0
mprotect(0x7f8463ff2000, 4096, PROT_READ) = 0
mprotect(0x646000, 4096, PROT_READ)     = 0
mprotect(0x7f8464222000, 4096, PROT_READ) = 0
munmap(0x7f8464387000, 34552)           = 0
socket(PF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, 0) = 3
setsockopt(3, SOL_SOCKET, SO_SNDBUF, [32768], 4) = -1 EINVAL (Invalid argument)
dup(2)                                  = 4
fcntl(4, F_GETFL)                       = 0x2 (flags O_RDWR)
brk(0)                                  = 0xe6b000
brk(0xe8c000)                           = 0xe8c000
fstat(4, {st_mode=S_IFCHR|0600, st_rdev=makedev(4, 1), ...}) = 0
ioctl(4, SNDCTL_TMR_TIMEBASE or SNDRV_TIMER_IOCTL_NEXT_DEVICE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f8464360000
lseek(4, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
write(4, "SO_SNDBUF: Invalid argument\n", 28SO_SNDBUF: Invalid argument
) = 28
close(4)                                = 0
munmap(0x7f8464360000, 4096)            = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by Efreak** [comment]
As mentioned, Aria2c has an issue with netlink as well. This is on both a current (build 14393.222) of running Ubuntu 16.04.1/xenial, and an insider (build 14926.1000) running trusty. Strace from insider/trusty is [here](https://gist.github.com/06e43e8fe00c74fe9fcd15451ebb2b2a); strace from stable/xenial is [here](https://gist.github.com/fe35f9b874fdc373351277057a464730)

**Comment by sunilmut** [comment]
@JasonLinMS as FYI. Thanks for the additional information @Efreak.

**Comment by coty-crg** [comment]
+1 would love to have this fixed

**Comment by sunilmut** [comment]
Just an update that 'ifconfig' and the 'ip' commands for network enumeration should be working in the latest fast insider build 14965, also see [release notes](https://msdn.microsoft.com/en-us/commandline/wsl/release_notes)

**Comment by isapir** [comment]
@sunilmut Any idea when build 14965 will be GA?  I see it on Insider Preview only ATM.  Thanks.

**Comment by benhillis** [comment]
The next update to Windows is the creators update coming early 2017.

**Comment by sunilmut** [comment]
@JasonLinMS as FYI

Just an update that the following networking commands should be working in the latest insider build (14971):
- ping
- ifconfig
- ip link
- ip addr show
- ifconfig
- whois
- nslookup (works but spews some errors about IP_RECVTOS)

Coming soon:
- ip route
- ip addr add

Not working:
- nmap (#1349)
- tracepath (#717)
- traceroute
- netstat
- ip addr add\delete

**Comment by n8cn** [comment]
This really cripples the ability to do "core network" function on a Windows machine.

When I heard (last year) that bash was coming to windows I said "Finally, I'll have a standard networking toolset available on my workstation - this is entirely the reason I carry either a Linux or Mac laptop for troubleshooting. 

Not having basic network functionality at a commandline pretty much shattered any hopes I had of getting down to "one workstation to rule them all."

someone bump me when this is fixed. Going back to OSX for a while..

**Comment by benhillis** [comment]
@n8cn - Thanks for the feedback.  Networking continues to be one of the areas that we're working to improve.  There's a ton of surface area and @sunilmut and others have been doing a great job chipping away at required functionality.

**Comment by sunilmut** [comment]
@n8cn - Thanks for the input. As you can see above, we have improved our networking tool significantly in the Insider build (which will be available in the `Creators Update`). And, more is coming.

**Comment by sunilmut** [comment]
Just FYI that in [15002](https://msdn.microsoft.com/en-us/commandline/wsl/release_notes), `ip addr add\del` & `ip route add\del` should be working (requires Windows admin).

**Comment by gpotter2** [comment]
For tcpdump/wireshark/some nmap functions, upvote
https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/17817685-support-af-packet-address-family-tcpdump-wireshar

**Comment by jsl303** [comment]
mtr still doesn't work on Build 15019. 
Start: Thu Feb  2 05:14:13 2017                                                                                         
HOST: DESKTOP Loss%   Snt   Last   Avg  Best  Wrst StDev                                                                 
Then nothing gets printed and exits to a prompt.

**Comment by Alysko** [comment]
No news about nmap support? Always the 'socket_bindtodevice: Protocol not available' error. Thanks.

**Comment by sunilmut** [comment]
@Alysko - Thanks for the input and the ping. We are looking into adding support for the socket `SO_BINDTODEVICE` socket option. But, I am not sure if that by itself will unblock `nmap` because it seems to have other dependencies such as the `PF_PACKET` socket family.

**Comment by powerpbx** [comment]
Would love to see `mtr` working.

**Comment by tommasov** [comment]
Well, ping works just fine.

**Comment by dswitzer** [comment]
I would definitely would love to use nmap and mtr. I use mtr often for troubleshooting. For Windows, I use WinMTR now but would prefer to stay at the command line.

**Comment by mmathys** [comment]
Bash on Windows is an extremely frustrating project

**Comment by JBFUK** [comment]
Hi, mtr and nmap still not working.  What a waste of time, may as well just run a Linux VM.

**Comment by gudenuff** [comment]
@mmathys 

Provided that it seamlessly runs many scripts allowing them to be executed right from PowerShell *without* having to run a separate VM, it's a god send.

Even as a normal user, I can run Unity and have Linux UI experience without bothering myself to install/configure a separate Linux machine.

If I was mostly using Linux, I'd use a dedicated station or a VM, but when I need POSIX environment just occasionally, BaSH on Windows is just was the doctor ordered.

**Comment by JBFUK** [comment]
@gudenuff until you find that quite a lot of commands and tools you want to use don't actually work, because, it's a half-baked implementation.

**Comment by gudenuff** [comment]
@JBFUK 

I haven't heard anyone from MSFT saying Bash on Windows provides complete Linux experience. It's like blaming Mercedes E63 that it doesn't do what you can have in McLaren. Just two separate worlds. But you can have some McLaren just occasionally where you really need it to.

Now call E63 the half-baked implementation of a sport car.

**Comment by DennisSc** [comment]
@JBFUK, nobody said nmap and mtr would work? Besides, Bash on WIndows is an implementation in progress as I understand it, so you'd better not expect "full-baked features" from dev / preview features

That being said, it is indeed still a long road; I recently failed with it while trying to cross-compile embedded linux kernels, due to chroot failure IIRC...

Nevertheless the idea is cool, keep up the work microsoft !

**Comment by JBFUK** [comment]
@gudenuff E63 absolutely is not a sports car.  It's an old-man's saloon, a fast one, but still an old man's saloon ;)

**Comment by DennisSc** [comment]
@JBFUK: yeah, I think that was his point - they're different, so you'd better not expect the same from both

**Comment by gudenuff** [comment]
@JBFUK

I don't live in a Linux world but I do need some tools from time to time. Just because, you know, there's a whopping universe of tools developed for POSIX.

I don't know if Bash would pay for itself from MSFT view, I don't know if it worth investment, but I know that many of those who don't need a full-fledged Linux experience and are good to go with *some* functionality right on Windows, those one, including myself, will be happy to be able to extend Windows automation experience.

Again, it's like blaming MSFT they haven't implemented UNIX socket functionality. What?

**Comment by JBFUK** [comment]
@DennisSc E63 is a pointless car, buy an uncomfortable high performance sports car, or, buy a comfortable saloon.  Maybe buy both.  Don't buy an uncomfortable high performance saloon that's disappointing on all fronts and costs more than the other two options combined...

Bringing it back on topic.  Either work with Windows based tools in Windows or work with Linux based tools in Linux.  It's a nice idea to be able to inter-operate but it clearly doesn't work.  Look at WINE under Linux - never really worked very well unless you just want to run Calc or Notepad.. waste of time.

Disappointing really, that's all I'm saying, disappointing.

**Comment by gudenuff** [comment]
@JBFUK 

Oh, man, get a life. Dump some of E63's MCU firmware to get some clue on what this saloon is capable of.

**Comment by DennisSc** [comment]
> ... an uncomfortable high performance saloon that's disappointing on all fronts and costs more ...

sounds like the average Mercedes Benz to me :octocat:

**Comment by mateusmedeiros** [comment]
@JBFUK Well, I can run ruby, node, erlang, elixir, nvim, nvim-qt, zsh, tmux, gcc, a lot of stuff, really. And by default when I download something from arch's repos, I expect it to work, because most of the time they work, so I'm sorry the tools you need are not working, but to say WSL is equivalent to having a Wine that can only run "calc" or "notepad" is a little unfair.

From what I've seen, the WSL team mostly follows the priorities on their uservoice page. I may be wrong, but I think what's blocking nmap right now is maybe https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/17817685-support-af-packet-address-family-tcpdump-wireshar ?

Anyway, I believe voting on a ticket or maybe opening a new one on that uservoice is your best bet.

**Comment by sunilmut** [comment]
The major blocker behind `mtr`, `tracepath`, `nmap` et. al. is the support for `AF_PACKET`, as correctly pointed out by @mateusmedeiros. Unfortunately, Windows itself does not has support for raw access to the ehternet. We (WSL team) continue to work with the Windows networking team to bring some of these features natively to Windows, which then can be lightened up in WSL.
And, yes, we do take the uservoice page seriously while prioritizing features. So, the best bet here is to head down to uservoice page and help us prioritize by casting your votes there.

**Comment by DJviolin** [comment]
You know you have to unsub from a git issue, when you reading about Mercedes E63s, and you still driving a 35 year old Audi 80...
And it still works because we replaced some stock parts in time, before some issues (although it's rotting like crazy).

**Comment by n8cn** [comment]
Yup. https://twitter.com/N8OUZ/status/865251213595619328

**Comment by JBFUK** [comment]
I guess you're right - unlucky that the tools I wanted to use don't work yet, maybe one day.  If the tools you like to work with will run then good for you.

**Comment by mmathys** [comment]
@gudenuff my comment sounds harsh sorry for that. I do appreciate the efforts of Microsoft and believe that they are doing the right thing with Windows.

However, one can't deny that at this state lots of things are broken and _**IMO**_ it's frustrating

**Comment by wildan2711** [comment]
I was dang excited when I saw this project because it means I didn't have to run a VM to do my networking tasks which was only available in Ubuntu, too bad it was the main weakness/issue in this project. Well just have to wait now, great job though WSL team..

**Comment by sunilmut** [comment]
I am going to close this issue out because it is a "catch-all" issue for a lot of different networking tools and has become unmanageable. It will be easier to track the individual networking tools that are not working with separate issues.

As of Creators Update, the following tools should work:

- ping
- dig
- ifconfig
- ip link
- ip addr show
- ifconfig
- whois
- nslookup
- ip route
- ip addr add/delete

Not working:
- nmap (#1349)
- tracepath (#717)
- traceroute (#1930)
- traceroute6 (#1979)

If you would like to see support for any additional tools, please open a new issue. Thanks all for chiming in here.

**Comment by yesterday679** [comment]
netstat still not work,
one question , how can i get 80 port when i use nginx ?

**Comment by ahmedAlmasri** [comment]
nmap still not working  

route_dst_netlink: cannot bind AF_NETLINK socket: Invalid argument

**Comment by wirehack7** [comment]
I am getting errors while doing `dig`:

```
$ dig google.de
../../../../lib/isc/unix/socket.c:2104: internal_send: 192.168.0.1#53: Invalid argument
../../../../lib/isc/unix/socket.c:2104: internal_send: 192.168.0.1#53: Invalid argument

; <<>> DiG 9.10.3-P4-Ubuntu <<>> google.de
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 57137
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.de.                     IN      A

;; ANSWER SECTION:
google.de.              299     IN      A       172.217.22.195

;; Query time: 108 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Sun Jul 23 14:10:02 DST 2017
;; MSG SIZE  rcvd: 54```

**Comment by sunilmut** [comment]
@wirehack7 - `dig` should work starting Insider build [16176](https://msdn.microsoft.com/en-us/commandline/wsl/release_notes#build-16176). It will also be available in Fall Creators Update.

**Comment by sunilmut** [comment]
@yesterday679 - Thanks for the feedback. We are aware of `netstat` not working. Regarding port 80, you will have to see where the collision is coming from, using netstat.exe. Or otherwise, just use a different port.

**Comment by sunilmut** [comment]
@ahmedAlmasri - Thanks for your post. `nmap` not working is a known issue and is being tracked at #1349 and #2039

**Comment by outofcontrolnate** [comment]
Trying to run tcpdump and getting "tcpdump: socket: Invalid argument" any suggestions or fixes?

**Comment by sunilmut** [comment]
@outofcontrolnate - `tcpdump` does not work currently because of the lack of `PF_PACKET` socket address family on WSL (and on Windows). [Here](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/17817685-support-af-packet-address-family-tcpdump-wireshar) is the user voice ticket. Please upvote if you would like us to prioritize this.

**Comment by ndeixonne** [comment]
Lol I was testing nmap and just fell on that topic. Some my friend fan of microsoft were telling me, no need for linux vm or double partition, bash on windows is exactly like a linux machine. Well, I guess this topic confirms my opinions >< Even if I understand it can be complicated to implement, networking can painful.

**Comment by haidescs** [comment]
How to solve this problem? @ldanicel  I meet the save issue with you, do you solve this issue yet?

**Comment by OvermindDL1** [comment]
@sunilmut Your link just displays `Forbidden` on an otherwise empty page...

And yep, still see the same issue near a year later...

**Comment by erlinux** [comment]
So is this problem solved?

**Comment by bitcrazed** [comment]
# **THIS THREAD IS CLOSED - please take discussion to more specific and therefore actionable issues.**

This bug is a pain because it lists several networking tools which were/are not working because of several different, and unrelated issues.

As noted in @sunilmut's [message just prior to closing this issue](https://github.com/Microsoft/WSL/issues/69#issuecomment-261668876):

* Ping was fixed in Creators Update. Nslookup/dig/etc. all work well on Fall Creators Update and later.
* We're working on improving support for several tools, inc. traceroute, route, etc. 

IP config tools are a lower priority right now, but can be worked-around by calling Windows commands to manage IP configuration via interop.

Tools requiring raw sockets (e.g. nmap, wireshark) are not currently supported on WSL. As noted in the [Microsoft docs on Raw Sockets](https://msdn.microsoft.com/en-us/library/windows/desktop/ms740548(v=vs.85).aspx), 

> Raw sockets offer the capability to manipulate the underlying transport, so they can be used for malicious purposes that pose a security threat. Therefore, only members of the Administrators group can create sockets of type SOCK_RAW on Windows 2000 and later. 

We are looking into improving things here, but it's not our highest priority since there are already many tools available on Windows for network penetration testing, packet capture, etc.

Please make this the last reply to this thread - take further discussion to more specific and therefore more actionable issues. 

Thanks.

**Comment by nakedgoat** [comment]
nmap has some errors and arp fails w/no such f/d on my windows box at home; however everything else works ping, dig, nslookup & ifconfig, on my work computer ALSO windows 10 also running bash as an admin (only real  difference is work computer is on a domain via vpn) @ work non of said commands work for me.. i've updated/upgraded both systems no real changes to much .. anyway hope this might help let me know if details are needed - even tho tread is closed :)

**Comment by adamfortuno** [comment]
tcptraceroute doesn't work either. 

```bash
tcptraceroute 10.140.103.197 3389
```

If I'm root, the error I get reads as follows:

> error opening device eth2: eth2: socket: Invalid argument

If I'm not root, the error I get reads as follows:

> error opening device wifi0: wifi0: socket: Invalid argument

I'm running the following version of tcptraceroute:
```
tcptraceroute 1.5beta7
Copyright (c) 2001-2006 Michael C. Toren <mct@toren.net>
Updates are available from http://michael.toren.net/code/tcptraceroute/
```

**Comment by MCMainiac** [comment]
When will this be fixed?

**Comment by abdilahrf** [comment]
2019 and still not fixed 👍

**Comment by jeyawn** [comment]
when can i use tcpdump?

**Comment by Tommasov** [comment]
Guys, this is not the right attitude! The missing features will be available when ready (if possible..) you should be thankful to the wls dev team for the great job done so far.

**Comment by nakedgoat** [comment]
What tude!?

**Comment by Tommasov** [comment]
Asking for features like in a restaurant... btw that was only my point of view. Peace

**Comment by Bruce-YC** [comment]
Have beening waiting for more than two years... Solution: Installed dual boot Arch last year 😂

**Comment by arrayown** [comment]
Wow, just came across this while trying to traceroute a host. It's been 3 years now 👎

**Comment by Karlheinzniebuhr** [comment]
Any updates?

**Comment by fordan** [comment]
This issue has been closed since Nov 2017. The fix is to use [WSL2](https://docs.microsoft.com/en-us/windows/wsl/wsl2-install) which has a Linux kernel and full system call functionality, although you need to be using Windows 10 build 18917 or higher which as of this date means a Windows Insider build (and opting in to the "skip ahead").

**Comment by fjleon1980** [comment]
ping is not working on 1809. this is embarrassing. why can't you fix something so basic?

**Comment by mr-ytseng** [comment]
It's not done yet 2020....

**Comment by RuiGuilherme** [comment]
1909 - 18363.628 | (2020-02-05) same issue. 

~ mtr 1.1.1.1
```
Failure to open IPv4 sockets: Permission denied
Failure to open IPv6 sockets: Permission denied
mtr: Failure to start mtr-packet: Invalid argument 
```

~ ping 1.1.1.1
`ping: socket: Operation not permitted`

~ tcptraceroute 1.1.1.1
`libnet_init() failed: libnet_open_raw4(): SOCK_RAW allocation failed: Permission denied `

~ nmap 192.168.0.151
```
Starting Nmap 7.70 ( https://nmap.org ) at 2020-02-05 13:19 -02                                                         Problem binding to interface , errno: 92 
socket_bindtodevice: Protocol not available
Problem binding to interface , errno: 92
```

**Comment by Tommasov** [comment]
> 1909 - 18363.628 | (2020-02-05) same issue.
> ~ mtr 1.1.1.1
> Failure to open IPv4 sockets: Permission denied
> Failure to open IPv6 sockets: Permission denied
> mtr: Failure to start mtr-packet: Invalid argument 
> 
> ~ ping 1.1.1.1
> ping: socket: Operation not permitted
> ~ tcptraceroute 1.1.1.1
> libnet_init() failed: libnet_open_raw4(): SOCK_RAW allocation failed: Permission denied 
> ~ nmap 192.168.0.151
> Starting Nmap 7.70 ( https://nmap.org ) at 2020-02-05 13:19 -02
> Problem binding to interface , errno: 92 
> socket_bindtodevice: Protocol not available
> Problem binding to interface , errno: 92

mts and nmap needs administrator privileges, open bash as admin and you can use it.
Ping should work even if the terminal is not elevated

**Comment by otcamargo** [comment]
I was having these problems on my debian distro when trying to ping any ip and when installing gems respectively:

```
ping: socket: Operation not permitted
``` 
``` 
Failure to open IPv4 sockets: Permission denied
Failure to open IPv6 sockets: Permission denied
```

This solution worked for me:

https://stackoverflow.com/questions/49800432/gem-cannot-access-rubygems-org/50349235#50349235

**Comment by rebcesp** [comment]
yet not fixed, march 2020

**Comment by dlakatos847** [comment]
WSL is just as it sounds: too good to be true. On OpenSuSE Leap 15.1, WSL2 root shell:
```
# mtr 1.1.1.1
Failure to open IPv4 sockets: Permission denied
Failure to open IPv6 sockets: Permission denied
mtr: Failure to start mtr-packet: Invalid argument
```

**Comment by fordan** [comment]
To repeat my earlier comment as a non-maintainer: This issue has been closed since Nov 2017. The fix is to use WSL2 which has a Linux kernel and full system call functionality, although you need to be using Windows 10 build 18917 or higher which as of this date means a Windows Insider build.

Even then you'll probably need to run mtr as root which is the cause of your "Permission Denied" errors since mtr modifies the packet TTL header to do its magic.

**Comment by fordan** [comment]
> WSL is just as it sounds: too good to be true. On OpenSuSE Leap 15.1, WSL2 root shell:

Works for me under Ubuntu.  I also just downloaded OpenSuSE Leap 15.1, changed it to WSL 2 and it worked fine for me as root. May want to do a wsl -l -v to verify your distro is set to use version 2.

**Comment by bitcrazed** [comment]
@dlakatos847 - from a Windows PowerShell / Cmd, could you run `wsl -l -v`?

I have a feeling your openSUSE may be running on WSL1

**Comment by dlakatos847** [comment]
> @dlakatos847 - from a Windows PowerShell / Cmd, could you run `wsl -l -v`?
> 
> I have a feeling your openSUSE may be running on WSL1

@bitcrazed You may be right, since the help is shown for `wsl -l -v`. Am I correct that Windows 10 2004 will bring WSL2 and a fix for this problem at the end of this month?

**Comment by dlakatos847** [comment]
I stand corrected, I unintentionally used WSL1. Windows Insider program, Windows Update and upgrading my Linux distro to WSL2 solved the problem. Thank you!

**Comment by bitcrazed** [comment]
As per the most passionately requests from the Linux community, we avoid automatically affecting your distros and apps - we don't upgrade your WSL distro instances, nor the binaries within them, and we do not auto-upgrade your WSL1 distros to WSL2 - that's something we leave to the user to perform via wsl.exe.

This MAY change in the future when WSL2 is provides >100% of WSL1's capabilities, but for now, while WSL1 and WSL2  have some different capabilities and behaviors, we leave it to you.

**Comment by bitcrazed** [comment]
![image](https://user-images.githubusercontent.com/961950/82683820-7c74fd00-9c06-11ea-9b62-bb5d4ab767ce.png)

> POLITE NOTE - please note that this issue is closed: if you have further issues with networking tech on WSL, please file a new issue with specific details about the issue you're seeing. Thanks.

**Comment by rebcesp** [comment]
> yet not fixed, march 2020

Hi, as everyone knows that the issue about using network tools in WSL has come up with problems especially with nmap but I have found something interesting. I have launched a trace while doing a TCP scan looking for versions to my router and I have seen that it has done it in a somewhat dirty way but I can only see it with `strace`


The syntax I have used is: **_nmap -sV -sT -sC 192.168.1.1_**

![image](https://user-images.githubusercontent.com/32257380/88815238-f62de780-d16f-11ea-8e9b-8af5d4c5dbdb.png)

I have not tried it with another IP, I also do not know why this happens and if I try to scan cleanly it does not work.

**Comment by ghzwael** [comment]
yet not fixed, September 2020

**Comment by bitcrazed** [comment]
@ghzwael Please note that, [as stated](https://github.com/microsoft/WSL/issues/69#issuecomment-632755683), this issue is closed.

We recommend that you adopt WSL2 (and convert any existing distro installations to WSL instances) which will provide you with much more compatibility and near-native performance.

**Comment by treysis** [comment]
@bitcrazed Is this a joke? IPv6 still not really working in WSL2...so, to use networking tools, I should upgrade to WSL2 which breaks networking all together? That's no solution.
